### ISSUE TITLE
IsAuthenticated is always true even if token is invalid

### ISSUE THREAD

**Issue by reft** [SpoonX/aurelia-authentication#339]
It seems that aurelia-authentication always assume that the LocalStorage token is valid if there is a key saved regardless of value?

If I manually modify the "aurelia_authentication" key in Local Storage and call auth.isAuthenticated I will always get back "true" even tho it's been tampered with and should return "false".

Can I validate the token on the clientside after logging in and receiving token from the server without a post to the server?

If I have an invalid key in my Local Storage, aurelia-authentication will always assume its valid and ignore the expire time resulting in redirects to my autenticated-startpage instead of the login-page. Users will be forced to manually delete the Local Storage token in order for them to be redirected to the login-page.

I encountered this problem after changing the secretkey on the server.

@RWOverdijk

**Comment by RWOverdijk** [comment]
@reft If the token can be decoded client side, it should be valid. If you modify the token and a request fails, that's on the user. It's not possible to verify the token client side because you need the secret.

**Comment by reft** [comment]
@RWOverdijk What if I use a middleware on the server to validate tokens in requests and then change the secretkey. Users logged in with the old secretkey will ofcourse be invalid and no longer authenticated serverside, but how would they get to the loginpage = auth:false instead of the startpage = auth:true? The clientside will still think the user is logged in.

**Comment by RWOverdijk** [comment]
@reft Well sure, if that's what you want to do go for it. :) But if you want to cater to people that change their own tokens in local storage... You're too kind to people that are messing around.

The next request made, if configured properly, will send the user to login. Let them fake being logged in to look at your client-side code, it's not a problem. If I can donload the code, I can see it. There's nothing you can do.

TL;DR; Client side permissions are **entirely** cosmetic and they do not matter. Make server-side secure, and client side work for the regular cases.

**Comment by reft** [comment]
@RWOverdijk I'm not sure I understand how to redirect the user to the login page if the clientside, (aurelia-authentication), still thinks the user is logged in. 

It doesn't matter if I try to redirect the user from the server to the login page because I'm using aurelia-routing. My loginpage is only available for users that are not authenticated. My startpage is only available for users that are authenticated, (using AuthenticateStep). 

If I change the secretkey or somehow the user has an invalid key in their localstorage **while they are logged in**, they will always be redirect to the authenticated startpage even tho they not authenticated on the server side:P

**Comment by RWOverdijk** [comment]
@reft If a user makes a request to the server, **any api call**, your server will return that the token is invalid, and they will be redirected to login. That's an automatic process.

```
// When authenticated, these endpoints will have the token added to the header of any requests (for authorization). Accepts an array of endpoint names. An empty string selects the default endpoint of aurelia-api.
configureEndpoints = null;
```

https://aurelia-authentication.spoonx.org/baseConfig.html

**Comment by reft** [comment]
@RWOverdijk If I modify the local storage key so the token is no longer valid and refresh the page I get 401 (Unauthorized) from all my requests as expected, (MVC), but aurelia-authentication doesn't understand that the token is invalid? It doesn't remove the token, redirect to the login page, logout or anything.. :S Im still authenticated on the client side. The server is returning 401 so why does it not redirect? Shouldn't aurelia-authentication logout if the server is returning 401? Could you give it a try and see if you get the same result?

**Comment by RWOverdijk** [comment]
Then don't change your local key.

But all joking aside, you can add your own interceptor for that. The current interceptor resolves when there's a 401, so you can catch that and simply check the response code. Every application handles these events differently and that's why it's not part of the default system. https://github.com/SpoonX/aurelia-authentication/blob/master/src/fetchClientConfig.js#L45,L77

What perhaps could be nice, is a new config option, something like `redirectOn401: bool`, which could then do the actual redirect there. Would that be something you'd like to do? It's as simple as adding the option to the config, implement the conditional in the code I just [linked to](https://github.com/SpoonX/aurelia-authentication/blob/master/src/fetchClientConfig.js#L45,L77) and importing the Redirect and then [using it](https://github.com/SpoonX/aurelia-authentication/blob/master/src/authenticateStep.js#L20). :)

**Comment by reft** [comment]
@RWOverdijk  Yes a new config option would be really nice :)

**Comment by RWOverdijk** [comment]
@reft Great :) So you'll give it a go in a PR?

**Comment by reft** [comment]
@RWOverdijk Im really swamped right now but it would be fun to do it in the future

**Comment by RWOverdijk** [comment]
@reft We're currently swamped as well, so unless someone else tackles this I think you can give it a go later :)

**Comment by reft** [comment]
Some more info - When manually changing the refresh token in local storage, aurelia auth gives:  _this8.authentication.updateAuthenticated is not a function
